core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -796,4 +803,244 @@ static Expression windowSelector(
         outputPhysType.record(expressions),
         parameter);
   }
+
+  /**
+   * Create enumerable implementation that applies sessionization to elements from the input
+   * enumerator based on a specified key. Elements are windowed into sessions separated by
+   * periods with no input for at least the duration specified by gap parameter.
+   */
+  public static Enumerable<Object[]> sessionize(Enumerator<Object[]> inputEnumerator,
+      int indexOfWatermarkedColumn, int indexOfKeyColumn, long gap) {
+    return new AbstractEnumerable<Object[]>() {
+      @Override public Enumerator<Object[]> enumerator() {
+        return new SessionizationEnumerator(inputEnumerator,
+            indexOfWatermarkedColumn, indexOfKeyColumn, gap);
+      }
+    };
+  }
+
+  private static class SessionizationEnumerator implements Enumerator<Object[]> {
+    private final Enumerator<Object[]> inputEnumerator;
+    private final int indexOfWatermarkedColumn;
+    private final int indexOfKeyColumn;
+    private final long gap;
+    private LinkedList<Object[]> list;
+    private boolean initialized;
+
+    SessionizationEnumerator(Enumerator<Object[]> inputEnumerator,
+        int indexOfWatermarkedColumn, int indexOfKeyColumn, long gap) {
+      this.inputEnumerator = inputEnumerator;
+      this.indexOfWatermarkedColumn = indexOfWatermarkedColumn;
+      this.indexOfKeyColumn = indexOfKeyColumn;
+      this.gap = gap;
+      list = new LinkedList<>();
+      initialized = false;
+    }
+
+    @Override public Object[] current() {
+      if (!initialized) {
+        initialize();
+        initialized = true;
+      }
+      return list.pollFirst();
+    }
+
+    @Override public boolean moveNext() {
+      return initialized ? list.size() > 0 : inputEnumerator.moveNext();
+    }
+
+    @Override public void reset() {
+      list.clear();
+      inputEnumerator.reset();
+      initialized = false;
+    }
+
+    @Override public void close() {
+      list.clear();
+      inputEnumerator.close();
+      initialized = false;
+    }
+
+    private void initialize() {
+      List<Object[]> elements = new ArrayList<>();
+      // initialize() will be called when inputEnumerator.moveNext() is true,
+      // thus firstly should take the current element.
+      elements.add(inputEnumerator.current());
+      // sessionization needs to see all data.
+      while (inputEnumerator.moveNext()) {
```

Comment:
Is this true ? Shouldn't we rely on the watermark and gap to decide if a session is expired ? What about the late arrive data ?

This implementation should not be blocking.

Reply:
It is true because this is batch only implementation.

For streaming cases, it needs to rely on watermark to know when it's ready to emit data. For batch cases, technically watermark always advances to the positive infinite at the beginning to mark all data has arrived (because it is bounded, or pre-known input data set). So this implementation is correct for batch case, because "watermark = positive infinity" means all data need to be processed.

I do have a plan to introduce watermark into calcite, and along with it, there should also be a test unbounded/streaming input source be built. By doing so we should be able to migrate current batch only implementation to batch-streaming unified implementation, which depends on watermark. However it will take some time to reach that point (with the support of community).

Reply:
I hear your concerns, but the operator did have a param named watermark.

Reply:
Sorry I might have missed something. What's the param that is named watermark? You mean the watermarked column, e.g. timestamp column specified by descriptor?

Reply:
Then why there are param named `indexOfWatermarkColumn` ?

Reply:
Watermark is a concept that is used to indicate the completeness of data, and it is not necessary an implementation.

The naming of "WatermarkedColumn" really means two things:
1. This is the column that provide event timestamp, which has the "event timestamp" semantic. (You can find "Event time vs. processing time" section in https://www.oreilly.com/radar/the-world-beyond-batch-streaming-101/ to know more about this semantic)
2. Because it has "event timestamp" semantic, we reply on "watermark" to know the data completeness and define late data. 

And windowing implemented by these series of PRs are coupled with "event timestamp" semantic, or we can just call it "event timestamp windowing" semantic.

What does it mean for batch and streaming sources? Say "hey,  this is the watermarked column, which gives the event timestamp. Let's apply windowing on this column. Wait a minute, how do I know the data is complete, thus I can close a window?"

For streaming source: "Hey it is streaming case, and we have watermark! If the watermark passes the end of window, you can mark the window is closed and all future data falls into that window is late".

For batch source: "Hey it is batch case, all data is known and there is no late data. Just close the window when all data are processed. And watermark also applies because when you ask the watermark, it always tell you that you can close the window". **In this case we can either implement watermark, which explicitly tell all window can be closed, or we just close them because this is batch. After all, it is just an implementation choice**, 

I think it is really a high-level argument and depends on if you accept this statement: batch is a special case of streaming. If you do accept this statement, using "watermark" is not wrong. Batch case can use all streaming terms (note that I am not saying using the same implementation). Also in the future, after streaming sources are built, either we change code here, or not, the behavior of what is build in this PR won't change cause it matches with "event timestamp windowing" semantic. So it is **backward compatible**.

Reply:
I do understand what a watermark is, no need to explain more, what i have concern upon is that the implementation has no relationship with a "watermark" definition but in the function doc, we represent it everywhere, which i think is not in consistent.

Reply:
Ok what's the recommendation then? Do you want "watermark" as a word removed?

Reply:
How about we add some notes in the function document that we only implements batch group window.

Reply:
Sure. Added a note for each windowing implementation to say they work for batch only.